File: BOJ/Dijkstra/11779-최소비용 구하기 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m, s, e;
    static ArrayList<int[]>[] graph;
    static int[] dist;
    static int[] prev;

    static ArrayList<Integer> r = new ArrayList<>();
    public static void Dijkstra(int v) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        dist[v] = 0;
        pq.add(new int[]{dist[v], v});

        while(!pq.isEmpty()) {
            int[] tmp = pq.poll();
            int d = tmp[0];
            int cur = tmp[1];

            if (d > dist[cur]) continue;
            for (int[] edge : graph[cur]) {
                int next = edge[0];
                int w = edge[1];
                int new_dist = w + d;
                if (new_dist < dist[next]) {
                    dist[next] = new_dist;
                    prev[next] = cur;
                    pq.add(new int[]{dist[next], next});
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        graph = new ArrayList[n + 1];
        dist = new int[n + 1];
        prev = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
            dist[i] = Integer.MAX_VALUE;
            prev[i] = -1;
        }

        for (int i = 0; i < m; i++) {
            int u, v, w;
            st = new StringTokenizer(br.readLine());
            u = Integer.parseInt(st.nextToken());
            v = Integer.parseInt(st.nextToken());
            w = Integer.parseInt(st.nextToken());

            graph[u].add(new int[]{v, w});
        }

        st = new StringTokenizer(br.readLine());
        s = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        Dijkstra(s);
        System.out.println(dist[e]);

        List<Integer> path = new ArrayList<>();
        int node = e;
        while (node != -1) {
            path.add(node);
            node = prev[node];
        }
        Collections.reverse(path);

        System.out.println(path.size());
        for (int city : path) {
            System.out.print(city + " ");
        }
    }
}

File: BOJ/Dijkstra/11779-최소비용 구하기 2.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <queue>
#include <algorithm>  // reverse() 함수를 위해 추가

using namespace std;

vector<vector<pair<int, int>>> graph;
vector<int> dist;
vector<int> prevCity;  // prev는 STL 함수와 혼동될 수 있으니 이름 변경 (예: prevCity)
vector<int> path;
int n, m;

void Dijkstra(int start) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    dist[start] = 0;
    pq.push({dist[start], start});

    while (!pq.empty()) {
        int d = pq.top().first;
        int cur = pq.top().second;
        pq.pop();

        if (d > dist[cur]) continue;

        for (auto edge : graph[cur]) {
            int next = edge.first;
            int w = edge.second;
            int new_dist = d + w;
            if (new_dist < dist[next]) {
                dist[next] = new_dist;
                prevCity[next] = cur;
                pq.push({dist[next], next});
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m;

    graph.resize(n + 1);
    dist.resize(n + 1, INT_MAX);
    prevCity.resize(n + 1, -1);

    for (int i = 0; i < m; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        graph[u].push_back({v, w});
    }

    int s, e;
    cin >> s >> e;

    Dijkstra(s);
    cout << dist[e] << endl;

    int cur = e;
    while (cur != -1) {
        path.push_back(cur);
        cur = prevCity[cur];
    }
    reverse(path.begin(), path.end());

    cout << path.size() << endl;
    for (int city : path) {
        cout << city << " ";
    }

    return 0;
}
